src/main/java/dan200/computercraft/shared/peripheral/speaker/SpeakerPeripheral.java

```diff
@@ -0,0 +1,243 @@
+/**
+ * This file is part of ComputerCraft - http://www.computercraft.info
+ * Copyright Daniel Ratcliffe, 2011-2016. Do not distribute without permission.
+ * Send enquiries to dratcliffe@gmail.com
+ */
+
+package dan200.computercraft.shared.peripheral.speaker;
+
+import dan200.computercraft.ComputerCraft;
+import dan200.computercraft.api.lua.ILuaContext;
+import dan200.computercraft.api.lua.LuaException;
+import dan200.computercraft.api.peripheral.IComputerAccess;
+import dan200.computercraft.api.peripheral.IPeripheral;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.SoundCategory;
+import net.minecraft.util.SoundEvent;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+public class SpeakerPeripheral implements IPeripheral {
+
+    private TileSpeaker m_speaker;
+    private long m_clock;
+    private long m_lastPlayTime;
+    private int m_notesThisTick;
+
+    public SpeakerPeripheral()
+    {
+        m_clock = 0;
+        m_lastPlayTime = 0;
+        m_notesThisTick = 0;
+    }
+
+    public SpeakerPeripheral(TileSpeaker speaker)
+    {
+        this();
+        m_speaker = speaker;
+    }
+
+    public void update() {
+        m_clock++;
+        m_notesThisTick = 0;
+    }
+
+    public World getWorld()
+    {
+        return m_speaker.getWorld();
+    }
+
+    public BlockPos getPos()
+    {
+        return m_speaker.getPos();
+    }
+
+    /* IPeripheral implementations */
+
+    @Override
+    public boolean equals(IPeripheral other)
+    {
+        if (other != null && other instanceof SpeakerPeripheral)
+        {
+            SpeakerPeripheral otherSpeaker = (SpeakerPeripheral) other;
+            return otherSpeaker.m_speaker == m_speaker;
+        }
+
+        else
+        {
+            return false;
+        }
+
+    }
+
+
+    @Override
+    public void attach(IComputerAccess computerAccess)
+    {
+    }
+
+    @Override
+    public void detach(IComputerAccess computerAccess)
+    {
+    }
+
+    @Override
+    public String getType()
+    {
+        return "speaker";
+    }
+
+    @Override
+    public String[] getMethodNames()
+    {
+        return new String[] {
+                "playSound", // Plays sound at resourceLocator
+                "playNote" // Plays note
+        };
+    }
+
+    @Override
+    public Object[] callMethod(IComputerAccess computerAccess, ILuaContext context, int methodIndex, Object[] args) throws LuaException
```

Comment:
So here's my issue with this PR: All the code called from this method is completely non-thread safe. Methods called from Lua always happen on the Lua thread, so it's not safe to call methods on anything managed by the main world directly unless it's your own data that you're correctly managing access to with syncronized blocks.

My recommendation is to use ILuaContext.issueMainThreadTask to ensure your sounds are played on the main thread. It's still safe and recommended to check all your parameters are correct before making this call so you can earlyout and give appropriate errors if they're incorrect.